staging/src/k8s.io/cloud-provider/controllers/service/controller.go

```diff
@@ -690,44 +714,69 @@ func (s *Controller) nodeSyncLoop() {
 		numServices-len(s.servicesToUpdate), numServices)
 
 	s.knownHosts = newHosts
 }
 
+// nodeSyncService syncs the nodes for one load balancer type service
+func (s *Controller) nodeSyncService(svc *v1.Service) bool {
+	if svc == nil || !wantsLoadBalancer(svc) {
+		return false
+	}
+	klog.V(4).Infof("nodeSyncService started for service %s/%s", svc.Namespace, svc.Name)
+	hosts, err := listWithPredicate(s.nodeLister, s.getNodeConditionPredicate())
+	if err != nil {
+		runtime.HandleError(fmt.Errorf("failed to retrieve node list: %v", err))
+		return true
+	}
+
+	if err := s.lockedUpdateLoadBalancerHosts(svc, hosts); err != nil {
+		runtime.HandleError(fmt.Errorf("failed to update load balancer hosts for service %s/%s: %v", svc.Namespace, svc.Name, err))
+		return true
+	}
+	klog.V(4).Infof("nodeSyncService finished successfully for service %s/%s", svc.Namespace, svc.Name)
+	return false
+}
+
 // updateLoadBalancerHosts updates all existing load balancers so that
-// they will match the list of hosts provided.
+// they will match the latest list of nodes with input number of workers.
 // Returns the list of services that couldn't be updated.
-func (s *Controller) updateLoadBalancerHosts(services []*v1.Service, hosts []*v1.Node) (servicesToRetry []*v1.Service) {
-	for _, service := range services {
-		func() {
-			if service == nil {
-				return
-			}
-			if err := s.lockedUpdateLoadBalancerHosts(service, hosts); err != nil {
-				runtime.HandleError(fmt.Errorf("failed to update load balancer hosts for service %s/%s: %v", service.Namespace, service.Name, err))
-				servicesToRetry = append(servicesToRetry, service)
-			}
-		}()
+func (s *Controller) updateLoadBalancerHosts(services []*v1.Service, workers int) (servicesToRetry []*v1.Service) {
+	if len(services) == 0 {
+		return
+	}
+	klog.V(4).Infof("Running updateLoadBalancerHosts(len(services)==%d, workers==%d)", len(services), workers)
+
+	// lock for servicesToRetry
+	lock := sync.Mutex{}
+	doWork := func(piece int) {
+		if shouldRetry := s.nodeSyncService(services[piece]); !shouldRetry {
+			return
+		}
+		lock.Lock()
+		defer lock.Unlock()
+		servicesToRetry = append(servicesToRetry, services[piece])
 	}
+
+	workqueue.ParallelizeUntil(context.TODO(), workers, len(services), doWork)
+	klog.V(4).Infof("Finished updateLoadBalancerHosts")
 	return servicesToRetry
 }
 
 // Updates the load balancer of a service, assuming we hold the mutex
 // associated with the service.
 func (s *Controller) lockedUpdateLoadBalancerHosts(service *v1.Service, hosts []*v1.Node) error {
-	if !wantsLoadBalancer(service) {
-		return nil
-	}
 	startTime := time.Now()
 	defer func() {
 		latency := time.Now().Sub(startTime).Seconds()
-		klog.V(4).Infof("It took %v seconds to update load balancer hosts for service %s/%s", latency, service.Namespace, service.Name)
+		klog.V(2).Infof("It took %v seconds to update load balancer hosts for service %s/%s", latency, service.Namespace, service.Name)
```

Comment:
Given that we have a metric for it, do we need level 2 logs?

Reply:
Yup. You are right. Change it back to V4